src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java

```diff
@@ -0,0 +1,92 @@
+import com.microsoft.azure.storage.blob.CloudBlob;
+import com.microsoft.azure.storage.blob.CloudBlobClient;
+import com.microsoft.azure.storage.blob.CloudBlobContainer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.StreamSupport;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+@Component
+public class BlobStorageRead {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BlobStorageRead.class);
+
+    private final CloudBlobClient cloudBlobClient;
+    private final Consumer<List<PDF>> pdfsConsumer;
+
+    @Autowired
+    public BlobStorageRead(CloudBlobClient cloudBlobClient, Consumer<List<PDF>> pdfsConsumer) {
+        this.cloudBlobClient = cloudBlobClient;
+        this.pdfsConsumer = pdfsConsumer;
+    }
+
+    @Scheduled(fixedDelayString = "${scan.interval}")
+    public void readBlobs() {
+        cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
+            try {
+                String containerName = cloudBlobContainer.getName();
+                CloudBlobContainer container = cloudBlobClient.getContainerReference(containerName);
+                readBlobsFromContainer(container);
+            } catch (URISyntaxException e) {
+                LOGGER.warn("Invalid URL", e);
+            } catch (StorageException e) {
+                LOGGER.warn("Could not obtain container reference", e);
+            }
+        });
+    }
+
+    private void readBlobsFromContainer(CloudBlobContainer container) {
+        StreamSupport.stream(container.listBlobs().spliterator(), false)
+            .filter(CloudBlob.class::isInstance)
+            .map(CloudBlob.class::cast)
+            .filter(cloudBlob -> cloudBlob.getName().toLowerCase().endsWith(".zip"))
+            // TODO: notify supplier about non-zip files
+            .map(cloudBlob -> {
+                try {
+                    BlobInputStream blobInputStream = container.getBlockBlobReference(cloudBlob.getName())
+                        .openInputStream();
+
+                    return processZipFile(blobInputStream);
+                } catch (StorageException e) {
+                    LOGGER.warn("Could not download data", e);
+                } catch (URISyntaxException e) {
+                    LOGGER.warn("Invalid URL", e);
+                }
+                return null;
+            })
+            .filter(Objects::nonNull)
+            .forEach(pdfsConsumer);
+    }
+
+    private List<PDF> processZipFile(BlobInputStream blobInputStream) {
+        List<PDF> pdfs = new ArrayList<>();
```

Comment:
Alternatively, you could use Stream.generate() for a more declarative approach. Up to you.

Reply:
In my opinion, the benefit would be simplicity. Not that important, though.

Reply:
Oh, and if those zips (or their whole decompressed content) turn out to be too large to be kept in memory, current way won't work, whereas the streamed one will allow for processing the PDFs one by one, without having to extract them all at once.

Reply:
Stream.generate() is an infinite stream and requires limiting. I have had a peek at `ZipFile.streams()` which utilises `Spliterators.spliterator()` and `ZipEntryIterator` where both of them require passing exact number of elements.

Of course, there is always `Stream.builder()` option, but it just another way for building the same array of elements. This brings another idea of converting `BlobInputStream` into `Iterator<PDF>`.

I am keen to hear your thoughts on this.

Reply:
We can just terminate the stream when there's nothing left.

Whether it's a stream or an iterator, doesn't matter. What's important is that we don't read the whole content of the zip file into memory if we don't have to. The agreed limit is already quite big (4*75 MB). If someone uploads a file that's too big to fit in memory, the job will keep failing as long as that file is there.

Reply:
I like this but it looks like premature optimisation. Let's address this in separate ticket.

Reply:
Not premature optimisation - we're reading potentially large files into memory and we can see how some input can cause this job to fail. I'm ok with addressing it in another pull request, though.